Feat: smart session guide v1 (#371)
reown-com/reown-docs: appkit/recipes/smart-sessions.mdx

---

> title: "How to use Smart Session with React"
> sidebarTitle: "Smart Sessions using AppKit"

Reviewer: ```suggestion
title: "How to integrate Smart Sessions with React"
sidebarTitle: "Smart Sessions using AppKit"
```

---

> Learn how to use Reown AppKit with Smart Sessions to grant permissions to delegate actions for a period of time.

Reviewer: ```suggestion
Learn how to use Reown AppKit with Smart Sessions to grant permission to delegate actions for a specific period of time.
```

Minor clarity fix

## Prerequisites
- A fundamental understanding of JavaScript and React.
- A minimal installation of AppKit in React.
- A new project Id obtained from Reown Cloud at https://cloud.reown.com
- A wallet with the private key in order to sign Transactions

## Final Project

> This is the completed working example to use and understand how smart sessions work with AppKit.

Reviewer: ```suggestion
Below is the complete working example that demonstrates how Smart Sessions work with AppKit.
```

<Card title="Appkit Example with Smart Session" icon="github" href="https://github.com/reown-com/appkit-web-examples/tree/main/react/react-wagmi-smart-session">
Clone this Github repository and follow the readme to try it out locally.
</Card>

## AppKit Minimal Installation

You can start a small project following the guidelines from our [installation React docs](https://docs.reown.com/appkit/react/core/installation#wagmi)

## Explain the functionality

> In this guide we are going to build the front-end from a dApp and a server to sign the transactions for the user.
> 
> Front-end
> - The user will connect using their email to our smart wallets and then grant the permission to the server related wallet in order to execute action in their behalf.
> - Check the configuration file in `src/config/configSmartSession.ts`.
> 
> API Server
> 
> The API server can have two endpoints:
> 1. getPublicKey: Return the publicKey for the keypair that is going to execute the different function in from the server. In order to achieve this, a private key needs to be added to the .env file.
> 
> 2. executeFunction: Trigger the function and checking the execution.
>     
> - Check the server configuration file in `config/index.ts`

Reviewer: ```suggestion
In this guide, we're going to build the front end of a dApp and a server to sign transactions on behalf of the user.

### Front-end

- The user will connect to AppKit's smart wallet using their email, and then grant permission to the server-linked wallet to execute actions on their behalf.
- Refer to the configuration file found in `src/config/configSmartSession.ts`.

### API Server

The API server can have two endpoints:
1. getPublicKey: Returns the publicKey for the keypair that is going to execute the different function in from the server. In order to achieve this, a private key needs to be added to the .env file.

2. executeFunction: Triggers the function and checking the execution.
    
- Refer to the server configuration file found in `config/index.ts`
```

Clone this Github repository and follow the readme to try it out locally.
</Card>

## AppKit Minimal Installation

You can start a small project following the guidelines from our [installation React docs](https://docs.reown.com/appkit/react/core/installation#wagmi)

## Explain the functionality

In this guide, we're going to build the front end of a dApp and a server to sign transactions on behalf of the user.

### Front-end

- The user will connect to AppKit's smart wallet using their email, and then grant permission to the server-linked wallet to execute actions on their behalf.
- Refer to the configuration file found in `src/config/configSmartSession.ts`.

### API Server

The API server can have two endpoints:
1. getPublicKey: Returns the publicKey for the keypair that is going to execute the different function in from the server. In order to achieve this, a private key needs to be added to the .env file.

2. executeFunction: Triggers the function and checking the execution.
    
- Refer to the server configuration file found in `config/index.ts`

<Frame height="100">
![Approve Transaction](/images/recipes/smartSession.png)
</Frame>

> 1. Frontend calls the endpoint `getPublicKey`
> 
> 2. The `requestPermission` modal appears for the user to approve using their wallet
> 
> 3. In the example, executing an action by calling the Server API (you can execute this action in several ways)
> 
> 4. `prepareCalls` RPC Call to the Blockchain API
> 
> 5. The server signs the hash with the private key
> 
> 6. `sendPrepareCalls` RPC Call to the Blockchain API
> 
> In order to simplify the code, we recommend that you use our util file from the server `util/prepareCalls.js`. All the RPC calls are performed to `https://rpc.walletconnect.org/v1/wallet?projectId=<PROJECT-ID>`

Reviewer: ```suggestion
1. Frontend calls the endpoint `getPublicKey`.

2. The `requestPermission` modal appears for the user to approve using their wallet.

3. In this example, the action is executed by calling the server API (though there are several ways to perform this action).

4. `prepareCalls` makes an RPC call to the Blockchain API.

5. The server signs the hash with the private key.

6. `sendPrepareCalls` makes an RPC call to the Blockchain API.

In order to simplify the code, it is recommended that you use the util file from the server; which can be found here - `util/prepareCalls.js`. All the RPC calls are performed to `https://rpc.walletconnect.org/v1/wallet?projectId=<PROJECT-ID>`.
```

</Frame>

### Part 2 | Execute the action

1. The server makes the call `wallet_prepareCalls`.

```js
 // make the prepare calls in our example
 // arg -> is the array of arguments to call the contract
const response = await makePrepareCalls(userAddress, chainId, contractAddress, ABI, functionName, context, arg);
```

2. The Result of the call should be signed with the private key.

```js
// Use the private key to sign the hash
const signature = await signatureCall(APPLICATION_PRIVATE_KEY, response.signatureRequest.hash);
```

3. Then we send the `wallet_sendPreparedCalls` with that signature.

```js
// send the prepared calls
const sendPreparedCallsResponse = await sendPreparedCalls({
    context: response.context,
    preparedCalls: response.preparedCalls,
    signature: signature,
});
```

4. The server can make several calls to `wallet_getCallsStatus` to check when it's confirmed.

```js
const userOpIdentifier = sendPreparedCallsResponse[0];

// function from our example to call wallet_getCallsStatus
const response = await getCallsStatus(userOpHash);
if (response.status === "CONFIRMED") { // check when the tx is confirmed
    return response;
}
```

> ## Common known issues
> 
> - "The method `wallet_grantPermissions` does not exist / is not available.’
>     - Please double check that your are using the right email adding +smart+session to end. Example: test@gmail.com -> test+smart+session@gmail.com
> 
> - Don't forget to add funds to the Server Wallet to pay the gas and to the smart wallet in case you need to deploy it.
> 
> - Sepolia network is now not working with smart sessions. Please use Sepolia Base

Reviewer: ```suggestion
## Common Known Issues

- "The method `wallet_grantPermissions` does not exist / is not available.’
    - Please double check that your are using the right email adding +smart+session to end. Example: test@gmail.com -> test+smart+session@gmail.com

- Don't forget to add funds to the **server wallet** to pay the gas fee and to the smart wallet in case you need to deploy it.

- Ethereum Sepolia isn’t currently working with Smart Sessions. Use Base Sepolia instead for now.
```
